### PR TITLE
Update types for relay-runtime 4.0 and react-relay 4.0

### PR DESCRIPTION
**Pull request by ckknight**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/relay/releases/tag/v4.0.0
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by ckknight** [comment]
We've been using these types within @bumped-inc for over a month now, and seems to be working fine for our purposes.

I converted the Flow definitions to TypeScript, and thus tried to match the Flow names as close as possible, rather than necessarily keeping the names of the relay 1.3 types (e.g. `OperationBase` => `OperationType`).

I'm also aware that Relay 5.0 released 3 days ago. Will likely spruce up these types again shortly.

File: types/react-relay/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for react-relay 1.3
+// Type definitions for react-relay 4.0
```
Comment (sibelius) [inline types/react-relay/index.d.ts]: can we move to react-relay 5.0.0?

**Review by sibelius — APPROVED** [review]

File: types/react-relay/index.d.ts (hunk 1 above)
Comment (ckknight) [reply types/react-relay/index.d.ts]: That is the plan, but I need to start using relay 5 and testing it. Might be able to get that all done today. Worst case, Monday.

Comment (sibelius) [reply types/react-relay/index.d.ts]: the only breaking change is internal api, won't change anything about types of public api

**Comment by alloy** [comment]
Great stuff! 👏

Will give this a try on our codebase on Monday.

**Comment by ckknight** [comment]
And here is my preliminary attempt at 5.0:

```patch
diff --git a/types/react-relay/index.d.ts b/types/react-relay/index.d.ts
index c14d6f7421..50d02a2dcf 100644
--- a/types/react-relay/index.d.ts
+++ b/types/react-relay/index.d.ts
@@ -1,4 +1,4 @@
-// Type definitions for react-relay 4.0
+// Type definitions for react-relay 5.0
 // Project: https://github.com/facebook/relay, https://facebook.github.io/relay
 // Definitions by: Johannes Schickling <https://github.com/graphcool>
 //                 Matt Martin <https://github.com/voxmatt>
diff --git a/types/relay-runtime/index.d.ts b/types/relay-runtime/index.d.ts
index f90228aae3..09e03fdb65 100644
--- a/types/relay-runtime/index.d.ts
+++ b/types/relay-runtime/index.d.ts
@@ -1,4 +1,4 @@
-// Type definitions for relay-runtime 4.0
+// Type definitions for relay-runtime 5.0
 // Project: https://github.com/facebook/relay, https://facebook.github.io/relay
 // Definitions by: Matt Martin <https://github.com/voxmatt>
 //                 Eloy Durán <https://github.com/alloy>
@@ -322,6 +322,8 @@ export type MissingFieldHandler =
           ) => ReadonlyArray<DataID | null | undefined> | null | undefined;
       };
 
+export const RelayDefaultMissingFieldHandlers: ReadonlyArray<MissingFieldHandler>;
+
 export interface OperationLoader {
     get(reference: unknown): NormalizationSplitOperation | null | undefined;
 
@@ -1111,6 +1113,8 @@ export function getStorageKey(
     field: NormalizationField | NormalizationHandle | ReaderField,
     variables: Variables
 ): string;
+export function getModuleComponentKey(documentName: string): string;
+export function getModuleOperationKey(documentName: string): string;
 
 // Declarative mutation API
 // ./mutations/RelayDeclarativeMutationConfig
@@ -1163,8 +1167,8 @@ export { RelayConnectionHandler as ConnectionHandler };
 
 // ./handlers/viewer/RelayViewerHandler
 interface RelayViewerHandler {
-    readonly VIEWER_ID: string;
-    update(store: RecordSourceProxy, payload: HandleFieldPayload): void;
+    readonly VIEWER_ID: DataID;
+    readonly VIEWER_TYPE: 'Viewer';
 }
 declare const RelayViewerHandler: RelayViewerHandler;
 export { RelayViewerHandler as ViewerHandler };
@@ -1242,7 +1246,10 @@ export const RelayProfiler: RelayProfiler;
 // Internal API
 export function deepFreeze<T extends object>(value: T): T;
 
-export const RelayFeatureFlags: {
-    MERGE_FETCH_AND_FRAGMENT_VARS: boolean;
-    PREFER_FRAGMENT_OWNER_OVER_CONTEXT: boolean;
-};
+// ./utils/RelayFeatureFlags
+
+interface FeatureFlags {
+    ENABLE_VARIABLE_CONNECTION_KEY: boolean;
+}
+
+export const RelayFeatureFlags: FeatureFlags;
diff --git a/types/relay-runtime/relay-runtime-tests.tsx b/types/relay-runtime/relay-runtime-tests.tsx
index e10c36fc4e..c1be5435d4 100644
--- a/types/relay-runtime/relay-runtime-tests.tsx
+++ b/types/relay-runtime/relay-runtime-tests.tsx
@@ -54,8 +54,9 @@ function handlerProvider(handle: any) {
         // Augment (or remove from) this list:
         case 'connection':
             return ConnectionHandler;
-        case 'viewer':
-            return ViewerHandler;
+        // case 'viewer':
+        //     // ViewerHandler is special-cased and does not have an `update` method
+        //     return ViewerHandler;
     }
     throw new Error(`handlerProvider: No handler provided for ${handle}`);
 }
```

**Comment by sibelius** [comment]
Can’t we have both v4 and v5?

**Comment by ckknight** [comment]
@sibelius, I assume we can. I was going to make a PR for 5.0 as soon as this one was reviewed by @alloy

**Comment by alloy** [comment]
Alright, going to test this out now.

**Comment by alloy** [comment]
Ok, I did a review and migration of [our codebase](https://github.com/artsy/emission/pull/1673). Made a few updates to the types, which I PR’ed on your fork https://github.com/ckknight/DefinitelyTyped/pull/1.

Otherwise this is looking great! 👌 ✨

**Review by alloy — CHANGES_REQUESTED** [review]
Ok, I did a review and migration of [our codebase](https://github.com/artsy/emission/pull/1673). Made a few updates to the types, which I PR’ed on your fork https://github.com/ckknight/DefinitelyTyped/pull/1.

Otherwise this is looking great! 👌 ✨

**Comment by alloy** [comment]
Do we really need to keep supporting v4 and v5 in parallel? Seems unnecessary to me, so perhaps we just do this release so that there is _a_ updated types package for v4 and then move on to v5 immediately afterwards?

**Comment by ckknight** [comment]
That was my idea, @alloy. Immediately move to v5.

**Comment by alloy** [comment]
One thing that I think we could start doing is copying over docs and inlining them, but that doesn’t need to block this PR at all.

**Review by alloy — APPROVED** [review]

**Comment by ckknight** [comment]
@alloy Yeah, that'd be a good idea.

**Comment by alloy** [comment]
Not going to wait any longer for the benchmark CI task, this is good as-is 👍 

Great job @ckknight! 👏

**Pull request merged**